src/components/views/dialogs/HostingSignupDialog.tsx

```diff
@@ -0,0 +1,158 @@
+
+interface IProps {
+    requestClose(): void,
+}
+
+interface IState {
+    completed: boolean,
+    error: string,
+}
+
+export default class HostingSignupDialog extends React.PureComponent<IProps, IState> {
+    iframeRef;
+    hostingSignupUrl: string;
+
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {
+            completed: false,
+            error: null,
+        };
+
+        this.iframeRef = React.createRef();
+        this.hostingSignupUrl = SdkConfig.get().hosting_signup.url;
+    }
+
+    private messageHandler = (message) => {
+        if (!this.hostingSignupUrl.startsWith(message.origin)) {
+            return;
+        }
+        switch (message.data.action) {
+            case 'account_credentials_request':
+                // noinspection JSIgnoredPromiseFromCall
+                this.sendAccountDetails();
+                break;
+            case 'setup_complete':
+                // Set as completed but let the user close the modal themselves
+                // so they have time to finish reading any information
+                this.setState({
+                    completed: true,
+                });
+                break;
+            case 'openid_credentials_request':
+                // noinspection JSIgnoredPromiseFromCall
+                this.fetchOpenIDToken();
+                break;
+            case 'close_dialog':
+                this.onFinished(true);
+                break;
+        }
+    }
+
+    private onFinished = (result: boolean) => {
+        if (result || this.state.completed) {
+            // We're done, close
+            this.props.requestClose();
+        } else {
+            const ConfirmDialog = sdk.getComponent('views.dialogs.ConfirmCloseHostingSignupDialog');
+            Modal.createDialog(
+                ConfirmDialog,
+                {
+                    onFinished: result => {
+                        if (result) {
+                            this.props.requestClose();
+                        }
+                    },
+                },
+            )
+        }
+    }
+
+    private sendMessage = (message) => {
+        this.iframeRef.contentWindow.postMessage(
+            message,
+            this.hostingSignupUrl,
+        )
+    }
+
+    private async fetchOpenIDToken() {
+        const token = await MatrixClientPeg.get().getOpenIdToken();
+        if (token && token.access_token) {
+            this.sendMessage({
+                action: 'openid_credentials',
+                tokenData: token,
+            });
+        } else {
+            this.setState({
+                error: _t("Failed to connect to your homeserver. Please close this dialog and try again."),
+            });
+        }
+    }
+
+    private async sendAccountDetails() {
+        this.sendMessage({
+            action: 'account_credentials',
+            credentials: {
+                accessToken: await MatrixClientPeg.get().getAccessToken(),
```

Comment:
uhh - why?

Reply:
Context, for anyone else reading this: EMS would use this to provide the user a possibility for an automated migration to the new Matrix host.